### DISCUSSION TITLE
Struggling to sync new table data

### DISCUSSION THREAD

**Discussion by badsyntax** [Mimetis/Dotmim.Sync#840] [Q&A]
Hi there. I've having a bit of a hard time understanding the correct process for handling new tables and syncing those tables.

(Note the client has already previously synced with the server.)

This is the process i've tried:

1. Create new table on server db using EF migrations
1. Add data into table on server db using EF migrations
1. Provision the server
1. Provision the client
1. Sync

As a result of the above process:

- The tracking table/triggers etc are created on the server db
- The table/tracking table/triggers etc are created on the client db
- No data is synced to the client db

Is there anythying I'm missing? I can't get the data in the newly created tables to sync. Thanks in advance for any advice.

This is the code i'm using:

```csharp
var serverProvider = new SqlSyncProvider("server-connection-string");
var clientProvider = new SqlSyncProvider("client-connection-string");

var setup = new SyncSetup(
    "ExistingTable",
    "NewTable"
);

Console.WriteLine("Will re-provision the server");
var remoteOrchestrator = new RemoteOrchestrator(serverProvider);
await remoteOrchestrator.ProvisionAsync(setup);

var agent = new SyncAgent(clientProvider, serverProvider);
agent.LocalOrchestrator.OnConflictingSetup(async args =>
{
    if (args.ServerScopeInfo != null)
    {
        Console.WriteLine("Will re-provision the client");
        args.ClientScopeInfo = await agent.LocalOrchestrator.ProvisionAsync(args.ServerScopeInfo, overwrite: true);
        args.Action = ConflictingSetupAction.Continue;
        return;
    }
    args.Action = ConflictingSetupAction.Abort;
});

Console.WriteLine("Will sync");

var result = await agent.SynchronizeAsync(setup);

Console.WriteLine(result);
```

**Comment by badsyntax** [comment]
It looks like the only way I can get this to work, is by `localOrchestrator.DeprovisionAsync(setup)`, then reprovision the client, then sync:

```csharp
var serverProvider = new SqlSyncProvider("server-connection-string");
var clientProvider = new SqlSyncProvider("client-connection-string");

var setup = new SyncSetup(
    "ExistingTable",
    "NewTable"
);

Console.WriteLine("Will re-provision the server");
var remoteOrchestrator = new RemoteOrchestrator(serverProvider);
await remoteOrchestrator.ProvisionAsync(setup);

var localOrchestrator = new LocalOrchestrator(clientProvider);
await localOrchestrator.DeprovisionAsync(setup);
await localOrchestrator.ProvisionAsync(await remoteOrchestrator.GetScopeInfoAsync());

var agent = new SyncAgent(clientProvider, serverProvider);
Console.WriteLine("Will sync");

var result = await agent.SynchronizeAsync(setup);

Console.WriteLine(result);
```

So my question is, if I create new tables on the server db, do I need to de-provision every client?

**Reply by Mimetis** [reply]
I think the 2 solutions you are using is fine.
If you see, at the end, that the new table is provisioned on server and client, well you did it well (new tracking table, new stored procedures / triggers)
And when you are synchronized, no datas are downloaded to your client, and ... that's normal too!

Let me explain: **DMS** works with a timestamp, stored on the client, that says basically (from the client perspective): 
- _"This timestamp stored locally is the last time I made a sync to the server"_
- _"Next time I shall do a sync, I will do from that timestamp to now"_
- _"Then I store a new timestamp for next sync"_

Your problem is that when you launch your second sync, basically you are syncing from last timestamp (call it N) to now.
And I guess you did not add / removed / edit any data in your new table since your provision time and now...

So far **DMS** is getting the data from that table that have never changed.

I'm pretty sure if you are adding a new record (or edit one) the record will be correctly downloaded.

So, what can you do?
- First solution, easiest: Once you have correctly provision client & server, the next client sync should be a `Reinitialize` (or `RenitializeWithUpload`)
- Second solution, tricky: You do the provision as you did, then you create a tmp scope for your new table only: A new scope has no timestamp, so it will download the whole table on the first sync. 
Make a sync with this new scope (only one table) to fully download this table locally
Now you can use your default scope to sync all the tables
A little bit more complex, but can resolve scenario where Renit is not an option

**Reply by badsyntax** [reply]
Hi. Thanks for the help.

For option 1 (Reinitialize):

After server and client are provisioned.

When I do `await agent.SynchronizeAsync(setup, SyncType.Reinitialize);`, I get the following error:

`Dotmim.Sync.SyncException: Seems you are trying another Setup tables that what is stored in your server scope database. Please create a new scope or deprovision and provision again your server scope.`

This error is unexpected as I've re-provisioned client and server, so the schema should match.

---

When I do: `await agent.SynchronizeAsync(SyncType.Reinitialize);`, I get the following error:

`Dotmim.Sync.SyncException: The DELETE statement conflicted with the REFERENCE constraint "FK_DefectComments_Defects_DefectId"`

I don't know why I get the delete error, but I assume because DMS is trying to delete the synced data, which has relationships with other entities? 

Any advice? I'll keep trying different things.

**Reply by Mimetis** [reply]
> When I do await agent.SynchronizeAsync(setup, SyncType.Reinitialize);, I get the following error:  Seems you are trying another Setup tables ...

Well, your `setup` instance (from `agent.SynchronizeAsync(setup, SyncType.Reinitialize)`) is not the same as the one saved in the `scope_info` table.
Check why and `ProvisionAsync` correctly your database, before making a sync (or create a new scope for this different setup)

To check what happens here, you can select the row from **scope_info** table and check the difference between your setup instance from your code and the column `sync_scope_setup` from your database. 
FYI, if you create the setup without a scope name, it's saved in the database with `sync_scope_name = 'DefaultScope'`

> When I do: await agent.SynchronizeAsync(SyncType.Reinitialize);, I get the following error: Dotmim.Sync.SyncException: The DELETE statement ...

Here, you are not passing any setup instance, then **DMS** get the instance from the `scope_info` table (that's why you do not have the first error raised)
The Reinitialize option will reinit the whole database. The first step is to delete all rows from all tables, then make a full sync.
If you have a **REFERENCE** error, that means the order of table deletes is not good.
The order in your `SyncSetup` instance is really important: 
Imagine you have **Product** and **ProductCateggory** tables, you can't create a **Product** if you don't have a **ProducCategory**: That means that your `SyncSetup` is `setup = new SyncSetup("ProductCategory", "Product")`
Table order matters.
If you have a circular reference or anything that prevents you to use a correct order, you still can use the [DisableConstraintsOnApplyChanges](https://dotmimsync.readthedocs.io/Configuration.html#disableconstraintsonapplychanges) option.

**Reply by badsyntax** [reply]
Thanks. It's good to know `SyncType.Reinitialize` will delete all rows, and that seems to be my problem here.

I've been able to successfully run `agent.SynchronizeAsync(setup, SyncType.Reinitialize)`, but only if there are no references to the synced data.

About the relationships, the constraint error is about the relationship between client data and sycned server data, not due to relationships within the server db.

This is our client database setup:

`VehiclesTable` (client db data)
`VehicleModelsTable` (sync server db data)

There's FK on `VehiclesTable` that points to `VehicleModelsTable`, so if we have any vehicles, we can't `SyncType.Reinitialize` as that would try to delete `VehicleModelsTable` rows, and will fail with a constraint error.

I will try the approach using a tmp scope to download the new table only, then try sync as normal. 

[edit] I'll also try [DisableConstraintsOnApplyChanges](https://dotmimsync.readthedocs.io/Configuration.html#disableconstraintsonapplychanges)

**Reply by badsyntax** [reply]
Hi. As I can't use `agent.SynchronizeAsync(setup, SyncType.Reinitialize)`, I've been able to sync the new table with data using a different scope for the new table. Thanks for the help.

**Reply by Mimetis** [reply]
nice !
